src/relations/db.py

```diff
@@ -0,0 +1,259 @@
+from pgconnstr import ConnectionString
+
+from utils import new_password
+
+logger = logging.getLogger(__name__)
+
+
+class DbProvides(Object):
+    """Defines functionality for the 'provides' side of the 'db' relation.
+
+    Hook events observed:
+        - relation-changed
+        - relation-departed
+        - relation-broken
+    """
+
+    def __init__(self, charm: CharmBase, admin: bool = False):
+        """Constructor for DbProvides object.
+
+        Args:
+            charm: the charm for which this relation is provided
+            admin: a boolean defining whether or not this relation has admin permissions, switching
+                between "db" and "db-admin" relations.
+        """
+        if admin:
+            self.relation_name = "db-admin"
+        else:
+            self.relation_name = "db"
+
+        super().__init__(charm, self.relation_name)
+
+        self.framework.observe(
+            charm.on[self.relation_name].relation_changed, self._on_relation_changed
+        )
+        self.framework.observe(
+            charm.on[self.relation_name].relation_departed, self._on_relation_departed
+        )
+        self.framework.observe(
+            charm.on[self.relation_name].relation_broken, self._on_relation_broken
+        )
+
+        self.admin = admin
+        self.charm = charm
+
+    def _on_relation_changed(self, event: RelationChangedEvent) -> None:
+        """Handle the legacy db/db-admin relation changed event.
+
+        Generate password and handle user and database creation for the related application.
+        """
+        # Check for some conditions before trying to access the PostgreSQL instance.
+        if (
+            "cluster_initialised" not in self.charm._peers.data[self.charm.app]
+            or not self.charm._patroni.member_started
+        ):
+            event.defer()
+            return
+
+        if not self.charm.unit.is_leader():
+            return
+
+        logger.warning(f"DEPRECATION WARNING - `{self.relation_name}` is a legacy interface")
+
+        unit_relation_databag = event.relation.data[self.charm.unit]
+        application_relation_databag = event.relation.data[self.charm.app]
+
+        # Do not allow apps requesting extensions to be installed.
+        if "extensions" in unit_relation_databag or "extensions" in application_relation_databag:
+            logger.error(
+                "ERROR - `extensions` cannot be requested through relations"
+                " - they should be installed through a database charm config in the future"
+            )
+            self.charm.unit.status = BlockedStatus("extensions requested through relation")
+            return
+
+        # Sometimes a relation changed event is triggered,
+        # and it doesn't have a database name in it.
+        database = event.relation.data[event.app].get("database")
+        if not database:
+            logger.warning("No database name provided")
+            event.defer()
+            return
```

Comment:
In what cases would a `defer()` resolve a missing `database` key from the relation data?
Do you expect it to happen eventually, set from the provider lib?